packages/core/src/shared/awsClientBuilderV3.ts

```diff
@@ -0,0 +1,152 @@
+interface AwsClient {
+    middlewareStack: {
+        add: MiddlewareStack<any, MetadataBearer>['add']
+    }
+}
+
+interface AwsClientOptions {
+    credentials: AwsCredentialIdentityProvider
+    region: string | Provider<string>
+    customUserAgent: UserAgent
+    requestHandler: {
+        metadata?: RequestHandlerMetadata
+        handle: (req: any, options?: any) => Promise<RequestHandlerOutput<any>>
+        destroy?: () => void
+    }
+    apiVersion: string
+    endpoint: string
+    retryStrategy: RetryStrategy | RetryStrategyV2
+}
+
+export class AWSClientBuilderV3 {
+    public constructor(private readonly context: AwsContext) {}
+
+    private getShim(): CredentialsShim {
+        const shim = this.context.credentialsShim
+        if (!shim) {
+            throw new Error('Toolkit is not logged-in.')
+        }
+        return shim
+    }
+
+    public async createAwsService<C extends AwsClient>(
+        type: AwsClientConstructor<C>,
+        options?: Partial<AwsClientOptions>,
+        region?: string,
+        userAgent: boolean = true,
+        settings?: DevSettings
+    ): Promise<C> {
+        const shim = this.getShim()
+        const opt = (options ?? {}) as AwsClientOptions
+
+        if (!opt.region && region) {
+            opt.region = region
+        }
+
+        if (!opt.customUserAgent && userAgent) {
+            opt.customUserAgent = [[getUserAgent({ includePlatform: true, includeClientId: true }), extensionVersion]]
+        }
+
+        if (!opt.retryStrategy) {
+            // Simple exponential backoff strategy as default.
+            opt.retryStrategy = new ConfiguredRetryStrategy(5, (attempt: number) => 1000 * 2 ** attempt)
+        }
+        // TODO: add tests for refresh logic.
+        opt.credentials = async () => {
+            const creds = await shim.get()
+            if (creds.expiration && creds.expiration.getTime() < Date.now()) {
+                return shim.refresh()
+            }
+            return creds
+        }
+
+        const service = new type(opt)
+        // TODO: add middleware for logging, telemetry, endpoints.
+        service.middlewareStack.add(telemetryMiddleware, { step: 'deserialize' } as DeserializeHandlerOptions)
+        return service
+    }
+}
+
+export function getServiceId(context: { clientName?: string; commandName?: string }): string {
+    return context.clientName?.toLowerCase().replace(/client$/, '') ?? 'unknown-service'
+}
+
+/**
+ * Record request IDs to the current context, potentially overriding the field if
+ * multiple API calls are made in the same context. We only do failures as successes are generally uninteresting and noisy.
+ */
+export function recordErrorTelemetry(err: Error, serviceName?: string) {
+    telemetry.record({
+        requestId: getRequestId(err),
+        requestServiceType: serviceName,
+        reasonDesc: getTelemetryReasonDesc(err),
+        reason: getTelemetryReason(err),
+        result: getTelemetryResult(err),
```

Comment:
The old `awsClientBuilder.ts` was reluctant to set these fields. But I think setting them here is low-risk and potentially adds useful info:

- if the caller is wrapped in `telemetry.xx.run()` then these fields will be overridden by the run() context finalizer anyway
- else, these fields give some useful info

todo: Maybe the `run()` finalizer logic should be smarter and check if there was a SDK failure (these fields)?